Fixes for using btree with the file system

w4kpm opened the pull request:
2 different patches -

1) when btree is running an does a sync, it ends up using mp_stream_posix_fsync which did a fsync by using a ioctl call, the problem was that the file ioctl didn't fulfill the flush request. 

2) add the ability to sync the db when using btree. This is merely a proxy for  Berkeley __bt_sync and exposes as db.sync() with no options -

w4kpm committed:
Allow file to respond to ioctl flush request

btree uses mp_stream_posix_fsync to flush the file. mp_stream_posix_fsync uses an ioctl, but the vfs_fat_fs didn't support it for flush.

I was having problems getting persistant btree on the filesystem without this.

NB:  when opening the file you should not use "w" for the file mode because it implicitly truncates the file. I have had more luck with using "rab" for the file mode when opening.

It would be nice to have a db flush exposed to python.

w4kpm committed:
Added a method to be able and sync the database

If you have longish operations on the db (such as logging data) it may be desireable to periodically sync the database to the disk. btree_sync merely exposes the berkley __bt_sync function to the user.

dpgeorge commented on extmod/vfs_fat_file.c:
```diff
@@ -140,6 +140,11 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(file_obj___exit___obj, 4, 4, file_obj
+    if (request == MP_STREAM_FLUSH) {
+	file_obj_flush(o_in);
+	return 0;
+    }
```
It would be better to inline (and simplify) file_obj_flush here, and then remove file_obj_flush and replace file_obj_flush_obj with mp_stream_flush_obj.  That would follow how the unix port does it.

Also, please follow the if-logic here and combine the if with the below if-else chain.

w4kpm committed:
Make changes per dpgeorge in review

It would be better to inline (and simplify) file_obj_flush here, and then remove file_obj_flush and replace file_obj_flush_obj with mp_stream_flush_obj. That would follow how the unix port does it.

Also, please follow the if-logic here and combine the if with the below if-else chain.

w4kpm commented:
Make changes that i think you had in mind - I'm not sure why I'm flagging an error on the modbtree.c commit . . .  Any suggestions?

Thanks,

w4kpm commented:
sorry - committed with an error - no coffee yet.

w4kpm committed:
continuation of last patch

w4kpm commented:
Error for modbtree.c was that the build took too long - not sure how to clear that -

dpgeorge commented:
Thanks.  I fixed up some formatting issues (eg replace tabs with spaces), rebased and merged in bd476acd41bcdd3718e73ec67ffc6a850d6aa453 and ec22d1739d256908f94f8c2c61a31b391f454fcf.

pfalcon commented:
Thanks for the patch. But it's good idea to use consistent method nomenclature for various objects. Python has file.flush(), so I renamed .sync() method from the patch to be btree.flush().
